src/ucx_uct_plugin.c

```diff
@@ -0,0 +1,1613 @@
+
+  status = uct_md_mem_dereg(comm->uct_iface->md, uct_memh->memh);
+  if (status != UCS_OK) {
+    WARN("Failed to deregister memh %p on comm %p: error %d", uct_memh, comm,
+         status);
+    return ncclSystemError;
+  }
+
+  uct_memh->comm = NULL;
+  free(uct_memh);
+  return ncclSuccess;
+}
+
+/* Outcome is either send_atp equal to 1 or 0 */
+static void nccl_uct_send_atp(nccl_uct_comm_t *comm, nccl_uct_rdesc_t *rdesc) {
+  ucs_status_t status;
+  nccl_uct_atp_t atp;
+  int i;
+
+  assert(rdesc->send_atp == 1);
+
+  status = uct_ep_fence(comm->uct_ep->ep, 0);
+  if (status != UCS_OK) {
+    return;
+  }
+
+  atp.id    = rdesc->desc.id;
+  atp.rdesc = rdesc->desc.peer_rdesc;
+  atp.count = rdesc->desc.count;
+
+  /* Sizes from isend() are lower or equal to their irecv() side */
+  for (i = 0; i < rdesc->desc.count; i++) {
+    atp.sizes[i] = rdesc->reqs[i].size;
+  }
+
+  status = uct_ep_am_short(comm->uct_ep->ep, NCCL_UCT_AM_ATP,
+                           (uint64_t)comm->remote_comm, &atp, sizeof(atp));
+  if (status == UCS_OK) {
+    rdesc->send_atp = 0;
+  }
+}
+
+static ncclResult_t nccl_uct_send(nccl_uct_comm_t *comm, void *data, int size,
+                                  nccl_uct_memh_t *uct_memh,
+                                  nccl_uct_rdesc_t *rdesc, int i,
+                                  void **request) {
+  ucs_status_t status;
+  uct_iov_t iov;
+  nccl_uct_req_t *req;
+
+  *request = NULL;
+
+  /* Details for local data */
+  iov.buffer = data;
+  iov.length = size;
+  iov.memh   = uct_memh->memh;
+  iov.stride = iov.length;
+  iov.count  = 1;
+
+  assert(size <= rdesc->desc.chunk[i].size);
+
+  req = nccl_uct_rdesc_get_req(rdesc, i, size); /* NCCL request */
+
+  status = uct_ep_put_zcopy(comm->uct_ep->ep, &iov, 1,
+                            (uint64_t)rdesc->desc.chunk[i].data,
+                            rdesc->desc.chunk[i].rkey, &req->completion);
+
+  if (status == UCS_OK) {
+    req->completion.count--;
+  } else if (status != UCS_INPROGRESS) {
+    return ncclSuccess;
+  }
+
+  rdesc->desc.chunk[i].matched = 1;
+  --rdesc->send_atp;
+
+  if (rdesc->send_atp == 1) {
+    nccl_uct_comm_rdesc_del(rdesc); /* all ->isend() were now matched */
+    nccl_uct_send_atp(comm, rdesc);
+  }
+
+  *request = req;
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_uct_isend(void *send_comm, void *data, int size,
+                                   int tag, void *mhandle, void **request) {
+  nccl_uct_comm_t *comm = send_comm;
+  nccl_uct_rdesc_t *rdesc;
+  int i;
+
+  *request = NULL;
+
+  for (rdesc = comm->rdesc_list.head; rdesc != NULL; rdesc = rdesc->next) {
+    for (i = 0; i < rdesc->desc.count; i++) {
+      if (rdesc->desc.chunk[i].matched || (rdesc->desc.chunk[i].tag != tag)) {
+        continue;
+      }
+
+      return nccl_uct_send(comm, data, size, mhandle, rdesc, i, request);
+    }
+  }
+
+  /* Progress here to make sure we receive non-solicited RTRs */
+  uct_worker_progress(comm->uct_worker->worker);
```

Comment:
yes, for completeness, in case we would only call `isend()`, it would fail because of missing rtr. it would then never call test, hence never progress, so never pull rtr and keep retrying.

Reply:
if progress is not enough, `->isend()` gets called again for retry and progress is called again.

the reasoning was that if we cannot send, it's because there is a lot of activity and other progresses are anyways incidentally progressing that worker. so in practice seemed that `while (uct_worker_progress()) {}` would only add "progress" load.

if there is only send traffic, in practice I guess it would be at startup, since right after traffic would start.

tbh i did not try using while(), it might not even be an issue.